torch_geometric/loader/cache.py

```diff
@@ -0,0 +1,55 @@
+from typing import Any, List, Tuple
+
+import torch
+from torch import Tensor
+from torch.utils.data import DataLoader
+
+BatchType = List[Tuple[Tensor, Tensor, int]]
+
+
+class CachedLoader:
+    r"""A caching loader intended to use in layer-wise inference
+    performed on GPU. During the first walk through the data
+    stored in :class:`torch.utils.data.DataLoader` batches are
+    transferred to GPU memory and cached.
+
+    Args:
+        loader (torch.utils.data.DataLoader): The wrapped data loader.
+        device (torch.device): The device to load the data to.
+    """
+    def __init__(self, loader: DataLoader, device: torch.device):
+        self.n_batches = len(loader)
+        self.iter = iter(loader)
+        self.cache_filled = False
+        self.device = device
+        self.cache = []
+
+    def cache_batch(self, batch: Any) -> BatchType:
+        n_id = batch.n_id.to(self.device)
+        if hasattr(batch, 'adj_t'):
+            edge_index = batch.adj_t.to(self.device)
+        else:
+            edge_index = batch.edge_index.to(self.device)
+        batch = (n_id, edge_index, batch.batch_size)
```

Comment:
Wondering why we don't cache the whole data object? Can we make this loader more general by just doing this?

Reply:
Done.

Reply:
@rusty1s Update, I have measured peak device memory usage:
```
Reddit:
Before: 8.36GB
Cache only n_id and edge_index: 9.4GB
Cache whole batch: 24.82GB

ogbn-products:
Before: 23.9GB
Cache only n_id and edge_index: 24.54GB
Cache whole batch: 62GB
```
As you can see, when caching the whole batch the peak memory usage jumps almost 3 times. No idea why memory usage is so high. Anyway, I have returned to the original implementation.